### PR TITLE
Dex: data loading base

### PR DESCRIPTION
**Pull request by 0xalecks**
- setup react query
- load wells (temporarily from .env file)
- load tokens from wells
- load token balances
- use caching

### CODE REVIEW

**Review by marshall-bean — CHANGES_REQUESTED** [review]
A few comments nothing too serious.

File: projects/dex-ui/src/components/App/OnLoad.tsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import { useAllTokensBalance } from "src/tokens/useTokenBalance";
+import { FC } from "src/types";
+import { useAccount, useNetwork } from "wagmi";
+
+export const OnLoad: FC<{}> = ({ children }) => {
+  const { address } = useAccount();
+  const { chain } = useNetwork();
+  // this call effectively acts as a 'prefetch' for the "get all token balances" query.
+  // we also refetch it when network or account changes
+  const { refetch } = useAllTokensBalance();
```
[the remark this answers is not in the document]
Comment (0xalecks) [reply projects/dex-ui/src/components/App/OnLoad.tsx:11]: Maybe I didn't write it clearly enough, but that won't work.

- The method invocation acts as the prefect. The result.refetch method is used later to literally re-fetch on network changes.

File: projects/sdk-wells/src/lib/tokens.ts [hunk 2]
```diff
@@ -74,7 +74,7 @@ export class Tokens {
    */
   findByAddress(address: string): Token | undefined {
     for (const token of this.tokens) {
-      if (token.address === address) return token;
+      if (token.address === address.toLowerCase()) return token;
```
[the remark this answers is not in the document]
Comment (0xalecks) [reply projects/sdk-wells/src/lib/tokens.ts:77]: No chance, we lowercase it when creating the Token

File: projects/dex-ui/src/tokens/useTokenBalance.tsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  }
+] as const;
+
+export const useAllTokensBalance = () => {
+  const tokens = useTokens();
+  const { address } = useAccount();
+  const queryClient = useQueryClient();
+
+  const tokensToLoad = Object.values(tokens);
+  if (tokensToLoad.length > 20) throw new Error("Too many tokens to load balances. Fix me");
```
[the remark this answers is not in the document]
Comment (0xalecks) [reply projects/dex-ui/src/tokens/useTokenBalance.tsx:27]: yeah this is temporary

File: projects/dex-ui/src/tokens/useTokenBalance.tsx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      })) as unknown as BigNumber[];
+      const balances: Record<string, TokenValue> = {};
+
+      for (let i = 0; i < res.length; i++) {
+        const value = res[i];
+        const token = tokensToLoad[i];
+        balances[token.symbol] = token.fromBlockchain(value);
+
+        // set the balance in the query cache too
+        queryClient.setQueryData(["token", "balance", token.symbol], { [token.symbol]: balances[token.symbol] });
```
[the remark this answers is not in the document]
Comment (marshall-bean) [reply projects/dex-ui/src/tokens/useTokenBalance.tsx:67]: Makes sense.

File: projects/dex-ui/src/wells/useWell.tsx [hunk 5]
```diff
@@ -1,18 +1,28 @@
...
-  });
+  const { data, isLoading, error } = useQuery<Well, Error>(
+    ["well", address],
+    async () => {
+      console.log("Fetching Well", address);
+      return sdk.wells.getWell(address);
+    },
+    {
+      placeholderData: () => {
+        const cachedWell = queryClient.getQueryData<Well[]>(["wells"])?.find((well) => well.address === address);
```
[the remark this answers is not in the document]
Comment (marshall-bean) [reply projects/dex-ui/src/wells/useWell.tsx:17]: Makes sense.

**Review by marshall-bean — APPROVED** [review]

**Pull request merged**